pallets/did/src/lib.rs

```diff
@@ -671,187 +765,208 @@ pub mod pallet {
+		#[pallet::weight(T::WeightInfo::submit_did_delete_operation())]
 		pub fn submit_did_delete_operation(
 			origin: OriginFor<T>,
-			did_deletion_operation: DidDeletionOperation<T::DidIdentifier>,
+			operation: DidDeletionOperation<T>,
 			signature: DidSignature,
 		) -> DispatchResultWithPostInfo {
-			// origin of the transaction needs to be a signed sender account
 			let sender = ensure_signed(origin)?;
 
-			let did_identifier = did_deletion_operation.get_did();
+			let did_identifier = operation.get_did();
 
-			// If specified DID does not exist, generate a DidNotPresent error.
 			let did_details = <Did<T>>::get(&did_identifier).ok_or(<Error<T>>::DidNotPresent)?;
 
-			// Verify the signature of the delete operation.
-			Self::verify_operation_signature_for_entry(&did_deletion_operation, &signature, &did_details)
-				.map_err(<Error<T>>::from)?;
-
-			// Verify that the operation counter is greater than the stored one
-			ensure!(
-				did_deletion_operation.tx_counter > did_details.last_tx_counter,
-				<Error<T>>::InvalidNonce
-			);
+			// Verify the signature and the nonce of the delete operation.
+			Self::verify_operation_validity_for_did(&operation, &signature, &did_details).map_err(<Error<T>>::from)?;
 
 			log::debug!("Deleting DID {:?}", did_identifier);
 			<Did<T>>::remove(&did_identifier);
 
 			Self::deposit_event(Event::DidDeleted(sender, did_identifier.clone()));
-			Ok(().into())
+
+			Ok(None.into())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	/// Verify the signature of a generic DidOperation, and
-	/// returns either Ok or a DidError. The paremeters are:
-	/// * op: a reference to the DID operation
+	/// Verify the validity (i.e., nonce and signature) of a generic
+	/// DidOperation and, if valid, update the DID state with the latest nonce.
+	///
+	/// * operation: the reference to the DID operation which validity is to be
+	///   verified
 	/// * signature: a reference to the signature
-	pub fn verify_did_operation_signature<O: DidOperation<T::DidIdentifier>>(
-		op: &O,
+	/// * did: the DID identifier to verify the operation signature for
+	pub fn verify_operation_validity_and_increase_did_nonce<O: DidOperation<T>>(
+		operation: &O,
 		signature: &DidSignature,
 	) -> Result<(), DidError> {
-		// Try to retrieve from the storage the details of the given DID. If there is no
-		// DID stored, generate a DidNotPresent error.
-		let did_entry: DidDetails =
-			<Did<T>>::get(op.get_did()).ok_or(DidError::StorageError(StorageError::DidNotPresent))?;
+		let mut did_details =
+			<Did<T>>::get(&operation.get_did()).ok_or(DidError::StorageError(StorageError::DidNotPresent))?;
+
+		Self::verify_operation_validity_for_did(operation, &signature, &did_details)?;
 
-		Self::verify_operation_signature_for_entry(op, signature, &did_entry)?;
+		// Update tx counter in DID details and save to DID pallet
+		did_details
+			.increase_tx_counter()
+			.expect("Increasing DID tx counter should be a safe operation.");
```

Comment:
if it's infallible it shouldn't return a result. If it's fallible we should return the error.